### PR TITLE
SDK DDB client update and search interface implementation

### PR DESCRIPTION
**Pull request by arjunkumargiri**
### Description
SDK DDB client update and search interface implementation. 
* Added implementation of DDB update
* Search requests are routed to search SDK client
* Update create and search to write as json attribute instead of a string under source attribute.
 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 1]
```diff
@@ -155,16 +200,140 @@ public CompletionStage<DeleteDataObjectResponse> deleteDataObjectAsync(DeleteDat
         }), executor);
     }
 
+    /**
+     * DDB data needs to be synced with opensearch cluster. {@link RemoteClusterIndicesClient} will then be used to
+     * search data in opensearch cluster.
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:205]: For my education, what's happening here?  

Are we writing to DDB and reading from OS for all "fetch" operations or are we only syncing up for search?

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 2]
```diff
@@ -155,16 +200,140 @@ public CompletionStage<DeleteDataObjectResponse> deleteDataObjectAsync(DeleteDat
...
-        return null;
+        return this.remoteClusterIndicesClient.searchDataObjectAsync(request, executor);
     }
 
     private String getTableName(String index) {
         // Table name will be same as index name. As DDB table name does not support dot(.)
         // it will be removed from name.
         return index.replaceAll("\\.", "");
     }
+
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:221]: All these static methods should probably be in a util class, in which case they might be public anyway.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public SdkClientModule() {
     SdkClientModule(String remoteMetadataType, String remoteMetadataEndpoint, String region) {
         this.remoteMetadataType = remoteMetadataType;
         this.remoteMetadataEndpoint = remoteMetadataEndpoint;
-        this.region = region;
+        this.region = region == null ? "us-west-2" : region;
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java:65]: Why us-west-2?  Should this be a configuration setting we can read from?

**Review by dbwiddis — COMMENTED** [review]
LGTM with some comments.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 2 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:221]: Good suggestion. Will update.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 1 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:205]: DDBOpenSearchClient will make use of DDB for CRUD operations. Data written to DDB will be synced with OpenSearch using offline integration between DDB and opensearch. SDKClient will then rely on opensearch only for search operations.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 4]
```diff
@@ -38,30 +42,41 @@
...
     private static final String DEFAULT_TENANT = "DEFAULT_TENANT";
 
     private static final String HASH_KEY = "tenant_id";
     private static final String RANGE_KEY = "id";
-    private static final String SOURCE = "source";
 
     private DynamoDbClient dynamoDbClient;
+    private RemoteClusterIndicesClient remoteClusterIndicesClient;
+
+    public DDBOpenSearchClient(DynamoDbClient dynamoDbClient, RemoteClusterIndicesClient remoteClusterIndicesClient) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:76]: Can we add java doc? Seems like `remoteClusterIndicesClient` will only be used for search operation? So if customer doesn't need the search functionality then they can just send null for `remoteClusterIndicesClient`?

File: plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java [hunk 5]
```diff
@@ -80,7 +80,8 @@ protected void configure() {
                 return;
             case AWS_DYNAMO_DB:
                 log.info("Using dynamo DB as metadata store");
-                bind(SdkClient.class).toInstance(new DDBOpenSearchClient(createDynamoDbClient()));
+                bind(SdkClient.class)
+                    .toInstance(new DDBOpenSearchClient(createDynamoDbClient(), new RemoteClusterIndicesClient(createOpenSearchClient())));
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java:84]: When you create a DDB client, do you see any sdk credential loading failures due to security issues? Last year when we tried to use aws sdk, the credential chain provider needs to read a file on the disk that is restricted access by OS so the SDK cannot be initialized. Wondering if this is a bug already fixed this year,

Comment (dbwiddis) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java:84]: @Zhangxunmt I was struggling with exactly that when trying to configure this client.

I was able to do a POC using AWS creds [here](https://github.com/opensearch-project/flow-framework/pull/631) and there were quite a few settings I had to add to [plugin-security.policy](https://github.com/opensearch-project/flow-framework/pull/631/files#diff-214b5d6612e51e2474aafa928918b67b571bfa31d6f4e398248840e5437714e9) and even that I needed to hard code a root path that I had permission to.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/util/JsonTransformer.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+
+public class JsonTransformer {
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    @VisibleForTesting
+    public static Map<String, AttributeValue> convertJsonObjectToItem(JsonNode jsonNode) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/util/JsonTransformer.java:30]: `convertJsonObjectToDDbAttributeMap`? as this method is only taking care of Dynamodb? Same goes for the other method name.

Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/util/JsonTransformer.java:30]: Rename the class too? `JsonTransformer` sounds too generic in this case.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 7]
```diff
@@ -76,16 +93,18 @@ public CompletionStage<PutDataObjectResponse> putDataObjectAsync(PutDataObjectRe
...
+                item.put(HASH_KEY, AttributeValue.builder().s(tenantId).build());
+                item.put(RANGE_KEY, AttributeValue.builder().s(id).build());
+                final PutItemRequest putItemRequest = PutItemRequest.builder().tableName(tableName).item(item).build();
 
-            dynamoDbClient.putItem(putItemRequest);
-            return new PutDataObjectResponse.Builder().id(id).created(true).build();
+                dynamoDbClient.putItem(putItemRequest);
+                return new PutDataObjectResponse.Builder().id(id).created(true).build();
+            } catch (IOException e) {
+                throw new OpenSearchStatusException("Failed to parse data object  " + request.id(), RestStatus.BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:106]: DDB would through runtime exception which will not be wrapped.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/util/JsonTransformer.java (hunk 6 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/util/JsonTransformer.java:30]: Will update method name, however would like to keep the name generic to support other json transformation functions in future

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**